lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.apache.lucene.index.Term;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.automaton.CompiledAutomaton;
+
+/**
+ * Interface to allow recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public interface QueryVisitor {
+
+  /**
+   * Called by leaf queries that match on a specific term
+   * @param query the leaf query visited
+   * @param term  the term the query will match on
+   */
+  void visitLeaf(Query query, Term term);
+
+  /**
+   * Called by leaf queries that do not match against the terms index
+   * @param query the leaf query visited
+   */
+  default void visitLeaf(Query query) {}
+
+  /**
+   * Called by leaf queries that match against a set of terms defined by a predicate
+   * @param query               the leaf query
+   * @param field               the field the query matches against
+   * @param predicateSupplier   a supplier for a predicate that will select matching terms
+   */
+  default void visitLeaf(Query query, String field, Supplier<Predicate<BytesRef>> predicateSupplier) {}
```

Comment:
I find this generalization odd. I think we should not add this without a user of this API. This is a general comment on this PR. It's hard to tell if it's sufficient if we don't have a clear user of the API? Should we move this to a branch and then iterate on it also with a user of it. ie. deprecate extractTerms and implement it with this?

Reply:
@s1monw I know it for sure, you were once a user of the original version of this APIon the Lucene issue - using InvocationVisitor to rewrite output of a QueryParser. I am still using this API for the same customer (with a slightly improved InvocationVisitor using MethodHandles and a better dispatch algorithm). I have to figure out what I can share about this customer, but the overlaps are there, so it's great to get this at some point.
What I like here is that this is hard-typed, but on the other hand I don't really like that visiting takes care of field names and terms: Not all queries are dealing with that. So it should be possible to have both ways: Just walk through a query tree in general, but if you are only interested in terms, it might be better to have it like proposed here. It should just be both possiblities (abstraction from terms and fields).

Reply:
I do remember what you are referring to. I think this can be a follow up to this change?

Reply:
If the visitor API works in general, adding more methods to improve usability (maybe using functional interfaces) would be a great win. I agree, can be done later.

The most important parts are covered here, so I am fine. I will still look at the code this evening to check if I find some problems. Most of the patch here deals with extractTerms, the actual changes in API are minor...